### ISSUE TITLE
No progress in training

### ISSUE THREAD

**Issue by tsmdt** [msiemens/HypheNN-de#6]
Hi, 

first of all, thanks for updating HypheNN-de to Tensorflow 2.13! 

I have difficulties in reproducing your training results (`Result: [0.0073155245152076989, 0.99240706430147785]`) and am not sure where to look for an error. 

Here are the steps I took for training:
1. I downloaded the latest wiktionary-dump from the provided link in the readme:  ([dewiktionary-latest-pages-articles.xml.bz2](https://dumps.wikimedia.org/dewiktionary/latest/dewiktionary-latest-pages-articles.xml.bz2) / 20-Jul-2023 17:03)
2. I ran `cargo run --release ../wiki-articles/dewiktionary-latest-pages-articles.xml ../wordlist.txt` which finished with no error and the following output: <br/>
```
cargo run --release ../wiki-articles/dewiktionary-latest-pages-articles.xml ../wordlist.txt
warning: use of deprecated trait `std::ascii::AsciiExt`: use inherent methods instead
 --> src/main.rs:7:17
  |
7 | use std::ascii::AsciiExt;
  |                 ^^^^^^^^
  |
  = note: `#[warn(deprecated)]` on by default

warning: unused import: `std::ascii::AsciiExt`
 --> src/main.rs:7:5
  |
7 | use std::ascii::AsciiExt;
  |     ^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

warning: `prepare_data` (bin "prepare_data") generated 2 warnings
    Finished release [optimized] target(s) in 0.03s
     Running `target/release/prepare_data ../wiki-articles/dewiktionary-latest-pages-articles.xml ../wordlist.txt`
Processed 1000 words
Processed 2000 words
Processed 3000 words
Processed 4000 words
Processed 5000 words
...
Processed 652000 words
Processed 653000 words
Processed 654000 words
Processed 655000 words
Randomizing line order...
Writing output file...
Done!
``` 
3. I started the training with `python train.py` and realized (if I am not mistaken) that you hardcoded a word limit for training in `dataset.py`, which is set to `TRAINING_SET = 150000`. That is a lot less than the available 655000 words generated through your Rust script. 
4. I therefore changed the 150000 limit to 90 % of the available training data (655000 words) in dataset.py: `TRAINING_SET = int(len(words) * 0.9)` which results in approx. 589500 words.
5. When starting the training through `python train.py` everything works fine, but the training itself doesn't make a lot of progress. Here is the training output:
```
python train.py                                
Building model...
2023-07-25 08:40:12.192699: I metal_plugin/src/device/metal_device.cc:1154] Metal device set to: Apple M1
2023-07-25 08:40:12.192721: I metal_plugin/src/device/metal_device.cc:296] systemMemory: 16.00 GB
2023-07-25 08:40:12.192725: I metal_plugin/src/device/metal_device.cc:313] maxCacheSize: 5.33 GB
2023-07-25 08:40:12.193053: I tensorflow/core/common_runtime/pluggable_device/pluggable_device_factory.cc:303] Could not identify NUMA node of platform GPU ID 0, defaulting to 0. Your kernel may not have been built with NUMA support.
2023-07-25 08:40:12.193339: I tensorflow/core/common_runtime/pluggable_device/pluggable_device_factory.cc:269] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 0 MB memory) -> physical PluggableDevice (device: 0, name: METAL, pci bus id: <undefined>)
Done

Preparing training data...
Processed 589500 entries (100 %)
Storing data...
Done

Training model...
Epoch 1/50
2023-07-25 08:40:24.986655: I tensorflow/core/grappler/optimizers/custom_graph_optimizer_registry.cc:114] Plugin optimizer for device_type GPU is enabled.
10489/10489 [==============================] - 44s 4ms/step - loss: 0.0511 - accuracy: 0.9386
Epoch 2/50
10489/10489 [==============================] - 46s 4ms/step - loss: 0.0493 - accuracy: 0.9416
Epoch 3/50
10489/10489 [==============================] - 45s 4ms/step - loss: 0.0494 - accuracy: 0.9417
Epoch 4/50
10489/10489 [==============================] - 45s 4ms/step - loss: 0.0496 - accuracy: 0.9416
...
Epoch 45/50
10489/10489 [==============================] - 44s 4ms/step - loss: 0.0532 - accuracy: 0.9419
Epoch 46/50
10489/10489 [==============================] - 44s 4ms/step - loss: 0.0533 - accuracy: 0.9419
Epoch 47/50
10489/10489 [==============================] - 43s 4ms/step - loss: 0.0533 - accuracy: 0.9420
Epoch 48/50
10489/10489 [==============================] - 44s 4ms/step - loss: 0.0537 - accuracy: 0.9418
Epoch 49/50
10489/10489 [==============================] - 44s 4ms/step - loss: 0.0537 - accuracy: 0.9419
Epoch 50/50
10489/10489 [==============================] - 44s 4ms/step - loss: 0.0536 - accuracy: 0.9420
Done
Time: 0:36:48.59
```

Maybe you are able to point me in the right direction, as I am in a bit of a loss here. I noticed that the training output in your readme works with a lot more data: `1253797/1253797`. Did you increase the training data limit while running this training?

Your model is also a lot more accurate: `0.99240706430147785` I am trying to reproduce your results (although you worked with an older wiktionary data set) and would be really happy, if you could help.

Thanks!

**Comment by msiemens** [comment]
Hey! I just checked with the latest dump and was able to train the model. I included your change to always use 90 % of the data set for training and fixes an issue with a word that contained the `*` character. I get these results:

```sh
# Training
Epoch 50/50
2449/2449 [==============================] - 2s 761us/step - loss: 0.0076 - accuracy: 0.9920

# Validation
Result: [0.009914549998939037, 0.9890826344490051]
Time: 0:00:15.89
```

Can you remove your model and dataset and try all the steps again? I really don't know what might cause the accuracy loss…

**Comment by tsmdt** [comment]
Hi,

I finally could reproduce your training results and identify the error on my end. My problem was running tensorflow on Mac M1 Pro with the `tensorflow-metal` package as recommended by an [Apple developers guide](https://developer.apple.com/metal/tensorflow-plugin/). With the package I could reproduce the problems I mentioned before, i.e. the accuracy capping at 94 %.

Without the `tensorflow-metal` package the training results and evaluation looks similar to yours:

```
python train.py                                                             
Building model...
Done

Preparing training data...
Processed 589500 entries (100 %)
Storing data...
Done

Training model...
Epoch 1/50
10484/10484 [==============================] - 9s 839us/step - loss: 0.0220 - accuracy: 0.9746
Epoch 2/50
10484/10484 [==============================] - 9s 836us/step - loss: 0.0113 - accuracy: 0.9870
Epoch 3/50
10484/10484 [==============================] - 9s 836us/step - loss: 0.0098 - accuracy: 0.9887
Epoch 4/50
10484/10484 [==============================] - 9s 837us/step - loss: 0.0091 - accuracy: 0.9896
Epoch 5/50
10484/10484 [==============================] - 9s 838us/step - loss: 0.0087 - accuracy: 0.9902
Epoch 6/50
10484/10484 [==============================] - 9s 837us/step - loss: 0.0083 - accuracy: 0.9906
Epoch 7/50
10484/10484 [==============================] - 9s 839us/step - loss: 0.0081 - accuracy: 0.9909
Epoch 8/50
10484/10484 [==============================] - 9s 838us/step - loss: 0.0079 - accuracy: 0.9911
Epoch 9/50
10484/10484 [==============================] - 9s 840us/step - loss: 0.0077 - accuracy: 0.9913
...
Epoch 45/50
10484/10484 [==============================] - 9s 832us/step - loss: 0.0063 - accuracy: 0.9933
Epoch 46/50
10484/10484 [==============================] - 9s 827us/step - loss: 0.0063 - accuracy: 0.9933
Epoch 47/50
10484/10484 [==============================] - 9s 860us/step - loss: 0.0063 - accuracy: 0.9933
Epoch 48/50
10484/10484 [==============================] - 9s 827us/step - loss: 0.0063 - accuracy: 0.9933
Epoch 49/50
10484/10484 [==============================] - 9s 856us/step - loss: 0.0063 - accuracy: 0.9933
Epoch 50/50
10484/10484 [==============================] - 9s 825us/step - loss: 0.0062 - accuracy: 0.9933
```

Here is the evaluation on the remaining 10 % of the wiki data set:
```
python validate.py                                               
Building model...
Done

Preparing validation data...
Processed 65500 entries (100 %)
Storing data...
Done

Validating model...
18715/18715 [==============================] - 7s 351us/step - loss: 0.0069 - accuracy: 0.9926

Done
Result: [0.006884512957185507, 0.9926362633705139]
Time: 0:00:08.30
```

So, thanks for helping! 🙂

**Comment by tsmdt** [comment]
I consider this issue fixed. 😉

**Comment by msiemens** [comment]
Interesting! I was able to reproduce the same behavior you described. 94 % accuracy in training, 72 % in validation. And interestingly enough, it seems like someone has reported a comparable issue on the Apple Developer Forums: https://developer.apple.com/forums/thread/734346, so we're not the only ones 🙂
